In order to initiliaze our program, we need to first import the necessary libraries related to our project.

In [ ]:
import tensorflow as tf
from tensorflow.keras import layers
import time

#Application 1

## Step 1 - Load the MNIST dataset in Tensorflow

In [ ]:
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


## Step 2 - Define the trainAndPredictMLP and model function


In [ ]:
def baseline_model(num_pixels, num_classes):

    #TODO - Application 1 - Step 6a - Initialize the sequential model
    model = tf.keras.models.Sequential()

    #TODO - Application 1 - Step 6b - Define a hidden dense layer with 8 neurons
    model.add(layers.Dense(8, input_dim=num_pixels,
                           kernel_initializer='normal', activation='relu'))

    #TODO - Application 1 - Step 6c - Define the output dense layer
    model.add(layers.Dense(num_classes, kernel_initializer='normal',
                           activation='softmax'))

    # TODO - Application 1 - Step 6d - Compile the model
    model.compile(loss='categorical_crossentropy', optimizer='adam',
                  metrics=['accuracy'])

    return model

In [ ]:
def trainAndPredictMLP(x_train, y_train, x_test, y_test):

    #TODO - Application 1 - Step 3 - Reshape the MNIST dataset - Transform the images to 1D vectors of floats (28x28 pixels  to  784 elements)
    num_pixels = x_train.shape[1] * x_train.shape[2]
    x_train = x_train.reshape((x_train.shape[0], num_pixels)).astype('float32')
    x_test = x_test.reshape((x_test.shape[0], num_pixels)).astype('float32')

    #TODO - Application 1 - Step 4 - Normalize the input values
    x_train = x_train / 255
    x_test = x_test / 255

    #TODO - Application 1 - Step 5 - Transform the classes labels into a binary matrix
    y_train = tf.keras.utils.to_categorical(y_train)
    y_test = tf.keras.utils.to_categorical(y_test)
    num_classes = y_test.shape[1]

    #TODO - Application 1 - Step 6 - Build the model architecture - Call the baseline_model function
    model = baseline_model(num_pixels,num_classes)

    #TODO - Application 1 - Step 7 - Train the model
    begin = time.time()
    model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=10,
              batch_size=200, verbose=2)
    end = time.time()
    total_time = end - begin
    #TODO - Application 1 - Step 8 - System evaluation - compute and display the prediction error
    scores = model.evaluate(x_test, y_test, verbose=0)
    print("Baseline Error: {:.2f}".format(100 - scores[1] * 100))
    print("Time to run: {:.2f}".format(total_time))
    return

In [ ]:
print("MLP - MNIST")
trainAndPredictMLP(x_train, y_train, x_test, y_test)

MLP - MNIST
Epoch 1/10
300/300 - 2s - 5ms/step - accuracy: 0.6627 - loss: 1.2493 - val_accuracy: 0.8307 - val_loss: 0.6306
Epoch 2/10
300/300 - 1s - 5ms/step - accuracy: 0.8587 - loss: 0.5161 - val_accuracy: 0.8804 - val_loss: 0.4283
Epoch 3/10
300/300 - 1s - 5ms/step - accuracy: 0.8877 - loss: 0.4054 - val_accuracy: 0.8957 - val_loss: 0.3728
Epoch 4/10
300/300 - 1s - 4ms/step - accuracy: 0.8995 - loss: 0.3611 - val_accuracy: 0.9065 - val_loss: 0.3324
Epoch 5/10
300/300 - 1s - 2ms/step - accuracy: 0.9079 - loss: 0.3274 - val_accuracy: 0.9117 - val_loss: 0.3109
Epoch 6/10
300/300 - 1s - 4ms/step - accuracy: 0.9128 - loss: 0.3094 - val_accuracy: 0.9159 - val_loss: 0.2988
Epoch 7/10
300/300 - 1s - 4ms/step - accuracy: 0.9161 - loss: 0.2976 - val_accuracy: 0.9165 - val_loss: 0.2883
Epoch 8/10
300/300 - 1s - 4ms/step - accuracy: 0.9179 - loss: 0.2891 - val_accuracy: 0.9186 - val_loss: 0.2823
Epoch 9/10
300/300 - 1s - 2ms/step - accuracy: 0.9204 - loss: 0.2821 - val_accuracy: 0.9214 - val_lo

## Complete Exercise 1:
Here I am "recreating" all the code above to facilitate how we can show the difference of neurons in the accuracy. Of course that in a real application this is not necessary (it is actually bad) but I will do this throughout the Lab for didactic reasons.

In [ ]:
import pandas as pd

def baseline_model(num_pixels, num_classes, neurons):
    # TODO - Application 1 - Step 6a: Initialize the sequential model
    model = tf.keras.models.Sequential()

    # TODO - Application 1 - Step 6b: Define a hidden dense layer with the specified number of neurons
    model.add(layers.Dense(neurons, input_dim=num_pixels, kernel_initializer='normal', activation='relu'))

    # TODO - Application 1 - Step 6c: Define the output dense layer
    model.add(layers.Dense(num_classes, kernel_initializer='normal', activation='softmax'))

    # TODO - Application 1 - Step 6d: Compile the model
    model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

    return model

def trainAndPredictMLP(x_train, y_train, x_test, y_test, neurons_list):
    # TODO - Application 1 - Step 3: Reshape the MNIST dataset - Transform the images to 1D vectors of floats
    num_pixels = x_train.shape[1] * x_train.shape[2]
    x_train = x_train.reshape((x_train.shape[0], num_pixels)).astype('float32') / 255
    x_test = x_test.reshape((x_test.shape[0], num_pixels)).astype('float32') / 255

    # TODO - Application 1 - Step 5: Transform the class labels into a binary matrix
    y_train = tf.keras.utils.to_categorical(y_train)
    y_test = tf.keras.utils.to_categorical(y_test)
    num_classes = y_test.shape[1]

    # Initialize a results list
    results = []

    for neurons in neurons_list:
        # TODO - Application 1 - Step 6: Build the model architecture - Call the baseline_model function
        model = baseline_model(num_pixels, num_classes, neurons)

        # TODO - Application 1 - Step 7: Train the model
        begin = time.time()
        history = model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=10, batch_size=200, verbose=0)
        end = time.time()

        # TODO - Application 1 - Step 8: System evaluation - compute and display the prediction error
        scores = model.evaluate(x_test, y_test, verbose=0)

        # Save results
        results.append({
            "Neurons": neurons,
            "Accuracy": scores[1] * 100,  # Convert to percentage
            "Training Time (s)": end - begin  # Total training time
        })

    # Convert results to a DataFrame
    results_df = pd.DataFrame(results)

    return results_df

# List of neurons to test
neurons_list = [8, 16, 32, 64, 128]

# Calling the function and get results
results_df = trainAndPredictMLP(x_train, y_train, x_test, y_test, neurons_list)

# Displaying the results
print(results_df)


   Neurons   Accuracy  Training Time (s)
0        8  91.979998          12.623777
1       16  94.059998          12.691282
2       32  95.810002          12.485858
3       64  96.890002          15.804418
4      128  97.570002          25.764613


## Likewise, here is the code for the Exercise 2, in which we see the diference in accuracy when chaning the Batch Sizes of the model.

In [ ]:
import pandas as pd

def baseline_model(num_pixels, num_classes, neurons=8):
    # TODO - Application 1 - Step 6a: Initialize the sequential model
    model = tf.keras.models.Sequential()

    # TODO - Application 1 - Step 6b: Define a hidden dense layer with 8 neurons
    model.add(layers.Dense(neurons, input_dim=num_pixels, kernel_initializer='normal', activation='relu'))

    # TODO - Application 1 - Step 6c: Define the output dense layer
    model.add(layers.Dense(num_classes, kernel_initializer='normal', activation='softmax'))

    # TODO - Application 1 - Step 6d: Compile the model
    model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

    return model

def trainAndPredictBatchSizes(x_train, y_train, x_test, y_test, batch_size_list):
    # TODO - Application 1 - Step 3: Reshape the MNIST dataset - Transform the images to 1D vectors of floats
    num_pixels = x_train.shape[1] * x_train.shape[2]
    x_train = x_train.reshape((x_train.shape[0], num_pixels)).astype('float32') / 255
    x_test = x_test.reshape((x_test.shape[0], num_pixels)).astype('float32') / 255

    # TODO - Application 1 - Step 5: Transform the class labels into a binary matrix
    y_train = tf.keras.utils.to_categorical(y_train)
    y_test = tf.keras.utils.to_categorical(y_test)
    num_classes = y_test.shape[1]

    # Initialize a results list
    results = []

    for batch_size in batch_size_list:
        # TODO - Application 1 - Step 6: Build the model architecture - Call the baseline_model function
        model = baseline_model(num_pixels, num_classes)

        # TODO - Application 1 - Step 7: Train the model with the given batch size
        begin = time.time()
        history = model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=10, batch_size=batch_size, verbose=0)
        end = time.time()

        # TODO - Application 1 - Step 8: System evaluation - compute and display the prediction error
        scores = model.evaluate(x_test, y_test, verbose=0)

        # Save results
        results.append({
            "Batch Size": batch_size,
            "Accuracy": scores[1] * 100,  # Convert to percentage
            "Training Time (s)": end - begin  # Total training time
        })

    # Convert results to a DataFrame
    results_df = pd.DataFrame(results)

    return results_df

# List of batch sizes to test
batch_size_list = [32, 64, 128, 256, 512]

# Calling the function and get results
results_df = trainAndPredictBatchSizes(x_train, y_train, x_test, y_test, batch_size_list)

# Displaying the results
print(results_df)


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


   Batch Size   Accuracy  Training Time (s)
0          32  92.350000          46.159856
1          64  92.830002          25.171666
2         128  92.159998          13.597790
3         256  92.210001          11.947284
4         512  91.229999           8.528251


## Exercise 3:
We will see the difference in changing the system performance from 'accuracy' to 'mse':

In [ ]:
import pandas as pd

def baseline_model_mse(num_pixels, num_classes, neurons=8):
    # TODO - Application 1 - Step 6a: Initialize the sequential model
    model = tf.keras.models.Sequential()

    # TODO - Application 1 - Step 6b: Define a hidden dense layer with 8 neurons
    model.add(layers.Dense(neurons, input_dim=num_pixels, kernel_initializer='normal', activation='relu'))

    # TODO - Application 1 - Step 6c: Define the output dense layer
    model.add(layers.Dense(num_classes, kernel_initializer='normal', activation='softmax'))

    # TODO - Application 1 - Step 6d: Compile the model with MSE as the metric
    model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['mse'])

    return model

def trainAndPredictMSE(x_train, y_train, x_test, y_test):
    # TODO - Application 1 - Step 3: Reshape the MNIST dataset - Transform the images to 1D vectors of floats
    num_pixels = x_train.shape[1] * x_train.shape[2]
    x_train = x_train.reshape((x_train.shape[0], num_pixels)).astype('float32') / 255
    x_test = x_test.reshape((x_test.shape[0], num_pixels)).astype('float32') / 255

    # TODO - Application 1 - Step 5: Transform the class labels into a binary matrix
    y_train = tf.keras.utils.to_categorical(y_train)
    y_test = tf.keras.utils.to_categorical(y_test)
    num_classes = y_test.shape[1]

    # TODO - Application 1 - Step 6: Build the model architecture - Call the baseline_model function
    model = baseline_model_mse(num_pixels, num_classes)

    # TODO - Application 1 - Step 7: Train the model
    begin = time.time()
    history = model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=10, batch_size=200, verbose=2)
    end = time.time()

    # TODO - Application 1 - Step 8: System evaluation - compute and display the MSE
    scores = model.evaluate(x_test, y_test, verbose=0)

    print(f"Mean Squared Error: {scores[1]:.4f}")
    print(f"Training Time: {end - begin:.2f} seconds")

    # Save results for analysis
    results = {
        "Metric": "MSE",
        "Value": scores[1],
        "Training Time (s)": end - begin
    }

    return results

# Calling the function and get results
results = trainAndPredictMSE(x_train, y_train, x_test, y_test)

# Displaying the results
print(pd.DataFrame([results]))


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/10
300/300 - 2s - 6ms/step - loss: 1.2186 - mse: 0.0534 - val_loss: 0.5975 - val_mse: 0.0274
Epoch 2/10
300/300 - 1s - 3ms/step - loss: 0.5124 - mse: 0.0236 - val_loss: 0.4329 - val_mse: 0.0196
Epoch 3/10
300/300 - 1s - 3ms/step - loss: 0.4190 - mse: 0.0189 - val_loss: 0.3799 - val_mse: 0.0170
Epoch 4/10
300/300 - 1s - 4ms/step - loss: 0.3684 - mse: 0.0164 - val_loss: 0.3338 - val_mse: 0.0147
Epoch 5/10
300/300 - 1s - 5ms/step - loss: 0.3266 - mse: 0.0144 - val_loss: 0.3088 - val_mse: 0.0135
Epoch 6/10
300/300 - 1s - 4ms/step - loss: 0.3086 - mse: 0.0135 - val_loss: 0.3006 - val_mse: 0.0132
Epoch 7/10
300/300 - 1s - 2ms/step - loss: 0.2979 - mse: 0.0130 - val_loss: 0.2927 - val_mse: 0.0128
Epoch 8/10
300/300 - 2s - 5ms/step - loss: 0.2905 - mse: 0.0126 - val_loss: 0.2889 - val_mse: 0.0126
Epoch 9/10
300/300 - 1s - 4ms/step - loss: 0.2849 - mse: 0.0123 - val_loss: 0.2827 - val_mse: 0.0123
Epoch 10/10
300/300 - 2s - 7ms/step - loss: 0.2794 - mse: 0.0121 - val_loss: 0.2826 - val_m

The accuracy-based run is better suited for scenarios where the primary objective is maximizing correct predictions.
The MSE-based evaluation, while computationally more demanding, is valuable when probability precision is critical, (in applications where confidence scores or probability calibration is required). Accuracy is better suited for this task as it directly reflects classification performance.

## Exercise 4:
We will train the model and save the associated weights in a file.

In [ ]:
def trainAndSaveModelWeights(x_train, y_train, x_test, y_test):
    # TODO - Application 1 - Step 3: Reshape the MNIST dataset - Transform the images to 1D vectors of floats
    num_pixels = x_train.shape[1] * x_train.shape[2]
    x_train = x_train.reshape((x_train.shape[0], num_pixels)).astype('float32') / 255
    x_test = x_test.reshape((x_test.shape[0], num_pixels)).astype('float32') / 255

    # TODO - Application 1 - Step 5: Transform the class labels into a binary matrix
    y_train = tf.keras.utils.to_categorical(y_train)
    y_test = tf.keras.utils.to_categorical(y_test)
    num_classes = y_test.shape[1]

    # TODO - Application 1 - Step 6: Build the model architecture - Call the baseline_model function
    model = baseline_model(num_pixels, num_classes, neurons=8)

    # TODO - Application 1 - Step 7: Train the model
    model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=10, batch_size=200, verbose=2)

    # TODO - Exercise 4: Save the weights of the trained model
    model.save_weights('mnist_model_weights.weights.h5')
    print("Model weights saved to 'mnist_model_weights.weights.h5'.")

    return model

# Train the model and save weights
model = trainAndSaveModelWeights(x_train, y_train, x_test, y_test)


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/10
300/300 - 2s - 7ms/step - accuracy: 0.7170 - loss: 1.1388 - val_accuracy: 0.8601 - val_loss: 0.5175
Epoch 2/10
300/300 - 1s - 2ms/step - accuracy: 0.8780 - loss: 0.4409 - val_accuracy: 0.8924 - val_loss: 0.3773
Epoch 3/10
300/300 - 1s - 4ms/step - accuracy: 0.8977 - loss: 0.3594 - val_accuracy: 0.9062 - val_loss: 0.3305
Epoch 4/10
300/300 - 1s - 4ms/step - accuracy: 0.9072 - loss: 0.3257 - val_accuracy: 0.9121 - val_loss: 0.3126
Epoch 5/10
300/300 - 1s - 2ms/step - accuracy: 0.9125 - loss: 0.3061 - val_accuracy: 0.9148 - val_loss: 0.2987
Epoch 6/10
300/300 - 1s - 3ms/step - accuracy: 0.9163 - loss: 0.2928 - val_accuracy: 0.9188 - val_loss: 0.2898
Epoch 7/10
300/300 - 1s - 3ms/step - accuracy: 0.9185 - loss: 0.2844 - val_accuracy: 0.9219 - val_loss: 0.2824
Epoch 8/10
300/300 - 1s - 4ms/step - accuracy: 0.9215 - loss: 0.2766 - val_accuracy: 0.9212 - val_loss: 0.2791
Epoch 9/10
300/300 - 1s - 3ms/step - accuracy: 0.9226 - loss: 0.2718 - val_accuracy: 0.9242 - val_loss: 0.2733
E

## Exercise 5,
We will load the saved weights and predict the first 5 images on the dataset:

In [ ]:
import numpy as np

def loadModelAndPredictFive(x_test, y_test, num_pixels, num_classes):
    # TODO - Application 1 - Step 6: Build the model architecture - Call the baseline_model function
    model = baseline_model(num_pixels, num_classes, neurons=8)

    # TODO - Exercise 4: Load the saved weights
    model.load_weights('mnist_model_weights.h5')  # Changed to correct file name.
    print("Model weights loaded from 'mnist_model_weights.h5'.")

    # Select the first 5 test images
    x_test_five = x_test[:5]
    y_test_five = y_test[:5]

    # TODO - Exercise 5: Make predictions on the first 5 test images
    predictions = model.predict(x_test_five)

    # Interpret predictions using np.argmax
    predicted_classes = np.argmax(predictions, axis=1)
    true_classes = np.argmax(y_test_five, axis=1)

    # Display results
    for i in range(5):
        print(f"Image {i + 1}: Predicted Class = {predicted_classes[i]}, True Class = {true_classes[i]}")

    return predicted_classes, true_classes

# Reshape test data to match input dimensions
num_pixels = x_train.shape[1] * x_train.shape[2]
x_test_reshaped = x_test.reshape((x_test.shape[0], num_pixels)).astype('float32') / 255
y_test_categorical = tf.keras.utils.to_categorical(y_test)
num_classes = y_test_categorical.shape[1]  # Define num_classes here

# Load the model and make predictions on the first 5 images
predicted_classes, true_classes = loadModelAndPredictFive(x_test_reshaped, y_test_categorical, num_pixels, num_classes)

FileNotFoundError: [Errno 2] Unable to synchronously open file (unable to open file: name = 'mnist_model_weights.h5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0)

#Application 2

## Step 1 - Train and predict on a CNN - Call the trainAndPredictCNN function

In [ ]:
trainAndPredictCNN(x_train,y_train,x_test,y_test) #We will only initialize this after the code of the function is complete

/usr/local/lib/python3.10/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
300/300 - 16s - 53ms/step - accuracy: 0.8918 - loss: 0.3893 - val_accuracy: 0.9511 - val_loss: 0.1677
Epoch 2/5
300/300 - 20s - 68ms/step - accuracy: 0.9521 - loss: 0.1619 - val_accuracy: 0.9698 - val_loss: 0.1051
Epoch 3/5
300/300 - 20s - 68ms/step - accuracy: 0.9642 - loss: 0.1182 - val_accuracy: 0.9766 - val_loss: 0.0797
Epoch 4/5
300/300 - 14s - 48ms/step - accuracy: 0.9717 - loss: 0.0930 - val_accuracy: 0.9799 - val_loss: 0.0668
Epoch 5/5
300/300 - 21s - 70ms/step - accuracy: 0.9762 - loss: 0.0778 - val_accuracy: 0.9814 - val_loss: 0.0591
Baseline Error: 1.86
Accuracy: 98.14%


As we can see, our baseline model is a much much better option than the previous MLP prediction.

## Step 2 - Define the trainAndPredictCNN and model function


In [ ]:
def CNN_model(input_shape, num_classes):
    # TODO - Application 2 - Step 5a - Initialize the sequential model
    model = tf.keras.Sequential()

    # TODO - Application 2 - Step 5b - Create the first hidden layer as a convolutional layer
    model.add(layers.Conv2D(filters=8, kernel_size=(3, 3), activation='relu', input_shape=input_shape))

    # TODO - Application 2 - Step 5c - Define the pooling layer
    model.add(layers.MaxPooling2D(pool_size=(2, 2)))

    # TODO - Application 2 - Step 5d - Define the Dropout layer
    model.add(layers.Dropout(rate=0.2))

    # TODO - Application 2 - Step 5e - Define the flatten layer
    model.add(layers.Flatten())

    # TODO - Application 2 - Step 5f - Define a dense layer of size 128
    model.add(layers.Dense(128, activation='relu'))

    # TODO - Application 2 - Step 5g - Define the output layer
    model.add(layers.Dense(num_classes, activation='softmax'))

    # TODO - Application 2 - Step 5h - Compile the model
    model.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])

    return model

In [ ]:
def trainAndPredictCNN(x_train, y_train, x_test, y_test):

    # TODO - Application 2 - Step 2 - reshape the data to be of size [samples][width][height][channels]
    x_train = x_train.reshape((x_train.shape[0], x_train.shape[1], x_train.shape[2], 1)).astype('float32')
    x_test = x_test.reshape((x_test.shape[0], x_test.shape[1], x_test.shape[2], 1)).astype('float32')

    # TODO - Application 2 - Step 3 - normalize the input values from 0-255 to 0-1
    x_train = x_train / 255.0
    x_test = x_test / 255.0

    # TODO - Application 2 - Step 4 - One hot encoding - Transform the classes labels into a binary matrix
    y_train = tf.keras.utils.to_categorical(y_train, 10)
    y_test = tf.keras.utils.to_categorical(y_test, 10)

    # TODO - Application 2 - Step 5 - Call the CNN_model function
    model = CNN_model(input_shape=(28, 28, 1), num_classes=10)

    # TODO - Application 2 - Step 6 - Train the model
    # Train the model on the training data and validate using the test data
    model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=5, batch_size=200, verbose=2)

    # TODO - Application 2 - Step 7 - Final evaluation of the model - compute and display the prediction error
    scores = model.evaluate(x_test, y_test, verbose=0)
    print(f"Baseline Error: {100 - scores[1] * 100:.2f}")
    print(f"Accuracy: {scores[1] * 100:.2f}%")

    return


## Exercise 6: Modify the size of the feature map within the convolutional layer

To modify the kernel size, we need to make some modifications to the CNN model and the trainAndPredictCNN

In [ ]:
def CNN_model(input_shape, num_classes, kernel_size):
    # Initialize the sequential model
    model = tf.keras.Sequential()

    # First convolutional layer with variable kernel size
    model.add(layers.Conv2D(filters=8, kernel_size=kernel_size, activation='relu', input_shape=input_shape))

    # Pooling layer
    model.add(layers.MaxPooling2D(pool_size=(2, 2)))

    # Dropout layer
    model.add(layers.Dropout(rate=0.2))

    # Flatten layer
    model.add(layers.Flatten())

    # Dense layer
    model.add(layers.Dense(128, activation='relu'))

    # Output layer
    model.add(layers.Dense(num_classes, activation='softmax'))

    # Compile the model
    model.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])

    return model
def trainAndPredictCNN_with_kernels(x_train, y_train, x_test, y_test, kernel_size):
    # Reshape the data
    x_train = x_train.reshape((x_train.shape[0], x_train.shape[1], x_train.shape[2], 1)).astype('float32')
    x_test = x_test.reshape((x_test.shape[0], x_test.shape[1], x_test.shape[2], 1)).astype('float32')

    # Normalize the input values
    x_train = x_train / 255.0
    x_test = x_test / 255.0

    # One hot encoding
    y_train = tf.keras.utils.to_categorical(y_train, 10)
    y_test = tf.keras.utils.to_categorical(y_test, 10)

    # Build the CNN model with the specified kernel size
    model = CNN_model(input_shape=(28, 28, 1), num_classes=10, kernel_size=kernel_size)

    # Train the model
    model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=5, batch_size=200, verbose=2)

    # Evaluate the model
    scores = model.evaluate(x_test, y_test, verbose=0)
    print(f"Kernel Size: {kernel_size} - Accuracy: {scores[1] * 100:.2f}%")

    return scores[1] * 100


And lastly, the code for the final change in kernel size

In [ ]:
def exercise_6(x_train, y_train, x_test, y_test):
    kernel_sizes = [(1, 1), (3, 3), (5, 5), (7, 7), (9, 9)]
    results = []

    for kernel_size in kernel_sizes:
        print(f"Testing with kernel size: {kernel_size}")
        accuracy = trainAndPredictCNN_with_kernels(x_train, y_train, x_test, y_test, kernel_size)
        results.append((kernel_size, accuracy))

    print("\nResults:")
    for kernel_size, accuracy in results:
        print(f"Kernel Size: {kernel_size} - Accuracy: {accuracy:.2f}%")

In [ ]:
exercise_6(x_train, y_train, x_test, y_test)

Testing with kernel size: (1, 1)
Epoch 1/5
300/300 - 11s - 38ms/step - accuracy: 0.8264 - loss: 0.6338 - val_accuracy: 0.9128 - val_loss: 0.3080
Epoch 2/5
300/300 - 20s - 68ms/step - accuracy: 0.8949 - loss: 0.3465 - val_accuracy: 0.9278 - val_loss: 0.2423
Epoch 3/5
300/300 - 9s - 29ms/step - accuracy: 0.9153 - loss: 0.2777 - val_accuracy: 0.9418 - val_loss: 0.1987
Epoch 4/5
300/300 - 11s - 36ms/step - accuracy: 0.9268 - loss: 0.2362 - val_accuracy: 0.9525 - val_loss: 0.1669
Epoch 5/5
300/300 - 11s - 37ms/step - accuracy: 0.9367 - loss: 0.2015 - val_accuracy: 0.9560 - val_loss: 0.1423
Kernel Size: (1, 1) - Accuracy: 95.60%
Testing with kernel size: (3, 3)
Epoch 1/5
300/300 - 16s - 52ms/step - accuracy: 0.8917 - loss: 0.3725 - val_accuracy: 0.9535 - val_loss: 0.1531
Epoch 2/5
300/300 - 20s - 67ms/step - accuracy: 0.9579 - loss: 0.1422 - val_accuracy: 0.9721 - val_loss: 0.0925
Epoch 3/5
300/300 - 21s - 69ms/step - accuracy: 0.9694 - loss: 0.1031 - val_accuracy: 0.9765 - val_loss: 0.0759


Smaller kernel sizes, such as 1x1 and 3x3, provided good accuracies (95.60% and 98.16%, respectively) while maintaining lower computational complexity. Larger kernel sizes, such as 5x5, 7x7, and 9x9, offered slightly higher accuracy (98.74% for 5x5, 98.87% for 7x7, and 98.73% for 9x9), demonstrating their ability to capture broader patterns in the images. However, the difference in accuracy among these larger kernels was very little. The 7x7 kernel size achieved the best accuracy (98.87%), making it ideal for this specific task. Overall, increasing the kernel size improved accuracy, but the gains did not improve a lot beyond 5x5, suggesting diminishing returns for larger kernels.

The fine details and broader patterns explanations were searched on the web, in order to learn what were some of the main differences.

## Exercise 7: Modify the number of neurons on the dense hidden layer

Likewise, we need to make some changes in the codes for the exercise 7

In [ ]:
def CNN_model(input_shape, num_classes, dense_neurons):
    # Initialize the sequential model
    model = tf.keras.Sequential()

    # Convolutional Layer
    model.add(layers.Conv2D(filters=8, kernel_size=(3, 3), activation='relu', input_shape=input_shape))

    # Pooling Layer
    model.add(layers.MaxPooling2D(pool_size=(2, 2)))

    # Dropout Layer
    model.add(layers.Dropout(rate=0.2))

    # Flatten Layer
    model.add(layers.Flatten())

    # Dense Layer with variable number of neurons
    model.add(layers.Dense(dense_neurons, activation='relu'))

    # Output Layer
    model.add(layers.Dense(num_classes, activation='softmax'))

    # Compile the model
    model.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])

    return model


def trainAndPredictCNN_with_neurons(x_train, y_train, x_test, y_test, dense_neurons):
    # Reshape the data
    x_train = x_train.reshape((x_train.shape[0], x_train.shape[1], x_train.shape[2], 1)).astype('float32')
    x_test = x_test.reshape((x_test.shape[0], x_test.shape[1], x_test.shape[2], 1)).astype('float32')

    # Normalize the input values
    x_train = x_train / 255.0
    x_test = x_test / 255.0

    # One hot encoding
    y_train = tf.keras.utils.to_categorical(y_train, 10)
    y_test = tf.keras.utils.to_categorical(y_test, 10)

    # Build the CNN model with the specified number of neurons
    model = CNN_model(input_shape=(28, 28, 1), num_classes=10, dense_neurons=dense_neurons)

    # Train the model
    start_time = time.time()
    model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=5, batch_size=200, verbose=2)
    end_time = time.time()

    # Evaluate the model
    scores = model.evaluate(x_test, y_test, verbose=0)
    accuracy = scores[1] * 100
    convergence_time = end_time - start_time

    print(f"Dense Neurons: {dense_neurons} - Accuracy: {accuracy:.2f}% - Time: {convergence_time:.2f}s")
    return accuracy, convergence_time


Thus, the final call for exercise 7 is:

In [ ]:
def exercise_7(x_train, y_train, x_test, y_test):
    neuron_sizes = [16, 64, 128, 256, 512]
    results = []

    for neurons in neuron_sizes:
        print(f"Testing with {neurons} neurons in the dense hidden layer")
        accuracy, convergence_time = trainAndPredictCNN_with_neurons(x_train, y_train, x_test, y_test, neurons)
        results.append((neurons, accuracy, convergence_time))

    print("\nResults:")
    for neurons, accuracy, time in results:
        print(f"Neurons: {neurons} - Accuracy: {accuracy:.2f}% - Convergence Time: {time:.2f}s")


In [ ]:
exercise_7(x_train, y_train, x_test, y_test)

Testing with 16 neurons in the dense hidden layer
Epoch 1/5
300/300 - 14s - 48ms/step - accuracy: 0.8074 - loss: 0.6554 - val_accuracy: 0.9260 - val_loss: 0.2629
Epoch 2/5
300/300 - 20s - 68ms/step - accuracy: 0.9253 - loss: 0.2569 - val_accuracy: 0.9475 - val_loss: 0.1805
Epoch 3/5
300/300 - 13s - 44ms/step - accuracy: 0.9419 - loss: 0.1988 - val_accuracy: 0.9576 - val_loss: 0.1451
Epoch 4/5
300/300 - 20s - 68ms/step - accuracy: 0.9510 - loss: 0.1660 - val_accuracy: 0.9640 - val_loss: 0.1238
Epoch 5/5
300/300 - 14s - 46ms/step - accuracy: 0.9571 - loss: 0.1439 - val_accuracy: 0.9661 - val_loss: 0.1089
Dense Neurons: 16 - Accuracy: 96.61% - Time: 89.26s
Testing with 64 neurons in the dense hidden layer
Epoch 1/5
300/300 - 16s - 52ms/step - accuracy: 0.8823 - loss: 0.4167 - val_accuracy: 0.9493 - val_loss: 0.1765
Epoch 2/5
300/300 - 20s - 67ms/step - accuracy: 0.9481 - loss: 0.1745 - val_accuracy: 0.9659 - val_loss: 0.1111
Epoch 3/5
300/300 - 14s - 46ms/step - accuracy: 0.9607 - loss: 0

Increasing the number of neurons in the dense hidden layer improved accuracy but also increased convergence time. The best balance was observed with 128 neurons, achieving an accuracy of 98.16% in 91.82 seconds. Adding more neurons, such as 256 or 512, slightly improved accuracy but with diminishing returns, while smaller layers, like 16 or 64 neurons, resulted in lower accuracy. Highlighting that 128 neurons is the optimal configuration for this task.

## Exercise 8: Modify the number of epochs used to train the model

In [ ]:
def trainAndPredictCNN_with_epochs(x_train, y_train, x_test, y_test, epochs):
    # Reshape the data
    x_train = x_train.reshape((x_train.shape[0], x_train.shape[1], x_train.shape[2], 1)).astype('float32')
    x_test = x_test.reshape((x_test.shape[0], x_test.shape[1], x_test.shape[2], 1)).astype('float32')

    # Normalize the input values
    x_train = x_train / 255.0
    x_test = x_test / 255.0

    # One hot encoding
    y_train = tf.keras.utils.to_categorical(y_train, 10)
    y_test = tf.keras.utils.to_categorical(y_test, 10)

    # Build the CNN model with 7x7 kernel size and 128 dense neurons
    model = CNN_model(input_shape=(28, 28, 1), num_classes=10, dense_neurons=128)

    # Train the model
    start_time = time.time()
    model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=epochs, batch_size=200, verbose=2)
    end_time = time.time()

    # Evaluate the model
    scores = model.evaluate(x_test, y_test, verbose=0)
    accuracy = scores[1] * 100
    convergence_time = end_time - start_time

    print(f"Epochs: {epochs} - Accuracy: {accuracy:.2f}% - Convergence Time: {convergence_time:.2f}s")
    return accuracy, convergence_time


Code to deploy the results of the change in epochs:

In [ ]:
def exercise_8(x_train, y_train, x_test, y_test):
    epoch_counts = [1, 2, 5, 10, 20]
    results = []

    for epochs in epoch_counts:
        print(f"Testing with {epochs} epochs")
        accuracy, convergence_time = trainAndPredictCNN_with_epochs(x_train, y_train, x_test, y_test, epochs)
        results.append((epochs, accuracy, convergence_time))

    print("\nResults:")
    for epochs, accuracy, time in results:
        print(f"Epochs: {epochs} - Accuracy: {accuracy:.2f}% - Convergence Time: {time:.2f}s")


In [ ]:
exercise_8(x_train, y_train, x_test, y_test)

Testing with 1 epochs
300/300 - 17s - 57ms/step - accuracy: 0.8908 - loss: 0.3929 - val_accuracy: 0.9580 - val_loss: 0.1522
Epochs: 1 - Accuracy: 95.80% - Convergence Time: 17.62s
Testing with 2 epochs
Epoch 1/2
300/300 - 16s - 54ms/step - accuracy: 0.8946 - loss: 0.3953 - val_accuracy: 0.9622 - val_loss: 0.1333
Epoch 2/2
300/300 - 21s - 71ms/step - accuracy: 0.9588 - loss: 0.1358 - val_accuracy: 0.9724 - val_loss: 0.0937
Epochs: 2 - Accuracy: 97.24% - Convergence Time: 42.34s
Testing with 5 epochs
Epoch 1/5
300/300 - 17s - 55ms/step - accuracy: 0.8935 - loss: 0.3834 - val_accuracy: 0.9523 - val_loss: 0.1612
Epoch 2/5
300/300 - 16s - 53ms/step - accuracy: 0.9557 - loss: 0.1509 - val_accuracy: 0.9699 - val_loss: 0.0980
Epoch 3/5
300/300 - 20s - 65ms/step - accuracy: 0.9686 - loss: 0.1058 - val_accuracy: 0.9782 - val_loss: 0.0742
Epoch 4/5
300/300 - 22s - 74ms/step - accuracy: 0.9745 - loss: 0.0831 - val_accuracy: 0.9799 - val_loss: 0.0632
Epoch 5/5
300/300 - 20s - 67ms/step - accuracy: 

Increasing the number of epochs improved accuracy but with diminishing returns and significantly longer convergence times. At 1 epoch, the accuracy was 95.80%, with a convergence time of 17.62 seconds. By 5 epochs, accuracy increased to 98.28%, achieving a good balance between accuracy and time (98.73 seconds). Training for 20 epochs resulted in the highest accuracy of 98.93%, but with a much longer convergence time of 374.69 seconds, highlighting that additional epochs beyond 5-10 provide minimal gains for the MNIST dataset while increasing computational cost.

## Exercise 9: Modify the CNN architecture with additional convolutional, max pooling layers and fully connected layers.

For the last exercise, we will modify the CNN architecture by adding extra convolutional, max pooling, and dense layers as specified and then evaluate the impact of this modified architecture on system accuracy.

The network will now include:

Convolutional Layer 1: 30 feature maps with kernel size 5x5.
Max Pooling Layer 1: Pool size 2x2.
Convolutional Layer 2: 15 feature maps with kernel size 3x3.
Max Pooling Layer 2: Pool size 2x2.
Dropout Layer: Drop 20% of neurons.
Flatten Layer: Convert 2D features to 1D.
Fully Connected Dense Layer 1: 128 neurons.
Fully Connected Dense Layer 2: 50 neurons.
Output Layer: 10 neurons for classification with softmax activation.

In [ ]:
def CNN_model_modified(input_shape, num_classes):
    # Initialize the sequential model
    model = tf.keras.Sequential()

    # First convolutional layer
    model.add(layers.Conv2D(filters=30, kernel_size=(5, 5), activation='relu', input_shape=input_shape))
    # First pooling layer
    model.add(layers.MaxPooling2D(pool_size=(2, 2)))

    # Second convolutional layer
    model.add(layers.Conv2D(filters=15, kernel_size=(3, 3), activation='relu'))
    # Second pooling layer
    model.add(layers.MaxPooling2D(pool_size=(2, 2)))

    # Dropout layer
    model.add(layers.Dropout(rate=0.2))

    # Flatten layer
    model.add(layers.Flatten())

    # Fully connected dense layer with 128 neurons
    model.add(layers.Dense(128, activation='relu'))

    # Fully connected dense layer with 50 neurons
    model.add(layers.Dense(50, activation='relu'))

    # Output layer
    model.add(layers.Dense(num_classes, activation='softmax'))

    # Compile the model
    model.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])

    return model

def trainAndPredictCNN_modified(x_train, y_train, x_test, y_test):
    # Reshape the data
    x_train = x_train.reshape((x_train.shape[0], x_train.shape[1], x_train.shape[2], 1)).astype('float32')
    x_test = x_test.reshape((x_test.shape[0], x_test.shape[1], x_test.shape[2], 1)).astype('float32')

    # Normalize the input values
    x_train = x_train / 255.0
    x_test = x_test / 255.0

    # One hot encoding
    y_train = tf.keras.utils.to_categorical(y_train, 10)
    y_test = tf.keras.utils.to_categorical(y_test, 10)

    # Build the modified CNN model
    model = CNN_model_modified(input_shape=(28, 28, 1), num_classes=10)

    # Train the model
    start_time = time.time()
    model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=5, batch_size=200, verbose=2)
    end_time = time.time()

    # Evaluate the model
    scores = model.evaluate(x_test, y_test, verbose=0)
    accuracy = scores[1] * 100
    convergence_time = end_time - start_time

    print(f"Modified CNN - Accuracy: {accuracy:.2f}% - Convergence Time: {convergence_time:.2f}s")
    return accuracy, convergence_time


In [ ]:
def exercise_9(x_train, y_train, x_test, y_test):
    print("Testing with Modified CNN Architecture")
    accuracy, convergence_time = trainAndPredictCNN_modified(x_train, y_train, x_test, y_test)
    print(f"\nResults:\nAccuracy: {accuracy:.2f}%\nConvergence Time: {convergence_time:.2f}s")


In [ ]:
exercise_9(x_train, y_train, x_test, y_test)

Testing with Modified CNN Architecture


/usr/local/lib/python3.10/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
300/300 - 40s - 133ms/step - accuracy: 0.8841 - loss: 0.3730 - val_accuracy: 0.9709 - val_loss: 0.0868
Epoch 2/5
300/300 - 40s - 133ms/step - accuracy: 0.9719 - loss: 0.0935 - val_accuracy: 0.9835 - val_loss: 0.0495
Epoch 3/5
300/300 - 41s - 137ms/step - accuracy: 0.9789 - loss: 0.0691 - val_accuracy: 0.9863 - val_loss: 0.0442
Epoch 4/5
300/300 - 40s - 135ms/step - accuracy: 0.9825 - loss: 0.0549 - val_accuracy: 0.9848 - val_loss: 0.0436
Epoch 5/5
300/300 - 35s - 118ms/step - accuracy: 0.9849 - loss: 0.0470 - val_accuracy: 0.9893 - val_loss: 0.0306
Modified CNN - Accuracy: 98.93% - Convergence Time: 202.90s

Results:
Accuracy: 98.93%
Convergence Time: 202.90s


The modified CNN architecture had an accuracy of 98.93% with a convergence time of 202.90 seconds. This architecture, whith the additional convolutional, pooling, and dense layers, improved accuracy slightly compared to simpler configurations, demonstrating its capability to capture more complex features in the MNIST dataset. However, the increased complexity resulted in a higher computational cost, making it suitable for tasks where maximum accuracy is prioritized over efficiency.